Adding extension methods to the generator

nzaugg opened the pull request:
Added methods for the string type, added null methods for nullable types and added another method to the DateTime extension methods

natezaugg committed:
Adding extension methods to the generator

legacybass commented on src/EFRepository.Generator/ExtensionMethodGenerator.cs:
```diff
@@ -1,237 +1,388 @@
+			};
+		})
+		.Where(c => c != null)
+		.Where(c =>
```
This formatting doesn't follow our standards of having the first item on the same line like for XAML and HTML, but it's a minor thing.

legacybass commented on src/EFRepository.Generator/ExtensionMethodGenerator.cs:
```diff
@@ -1,237 +1,388 @@
+				builder.AppendLine(CreateNullMethodFunctions(member.Name));
+
+				builder.AppendLine($@"
+	/// <summary>
```
These should be two tabs deep to match normal class formatting.

legacybass commented on src/EFRepository.Generator/ExtensionMethodGenerator.cs:
```diff
@@ -1,237 +1,388 @@
+	/// Filter the <see cref=""IQueryable""/> of {dbSetClass.Name} by {member.Name}
+	/// </summary>
+	/// <param name=""value"">The {type} which {member.Name} should be equal</param>
+	public static IQueryable<{dbSetClass.Name}> By{member.Name}(this IQueryable<{dbSetClass.Name}> query, {type}? value)
```
You're using the type variable here, but we know the type to be a string at this point. It probably doesn't matter, but was there a specific reason for this?

legacybass replied:
Could we use that same `nullable` variable here to determine if the string should be nullable? At minimum this can cause problems with the full framework, since strings are always nullable there. And this project still targets full framework which doesn't support the whole `EnableNullable` thing.

legacybass replied:
Now that I've seen the `CreateNullMethodFunctions` method, shouldn't we also use that for `string`? Those can be null as well, and we may want to filter by that.

nzaugg replied:
I tried to write the string functions in such a way that they can be nullable or not and we handle them appropriately. If they are null, a null check is performed thus resolving both the new and old styles.

nzaugg replied:
The string is a special case in which I want to use the "IsNullOrWhitespace" function instead of simply checking for null

legacybass replied:
My question was more about whether or not we wanted to have IsNull or IsNotNull extension methods for strings. Right now we don't have that, and we may want it. The user may want to both check that the string isn't null and check that it has certain characteristics. Right now, the code would just call the check for characteristics and return the query unchanged if the string is null.

nzaugg replied:
It is generating the IsNull and IsNotNull functions, but I think I'd like to add a couple more of it makes sense. By{field}IsNullOrWhiteSpace() and By{field}IsNotNullOrWhiteSpace()

Thoughts on that?

legacybass commented on src/EFRepository.Generator/ExtensionMethodGenerator.cs:
```diff
@@ -1,237 +1,388 @@
+using System.Linq;
+using Microsoft.EntityFrameworkCore;
+
+namespace {dbSetClass.ContainingNamespace};
```
We cannot use this feature here since this project also targets .NET Framework, which doesn't support the new namespace syntax. This also presupposes that the user is running version 10 of the language, which may also not be the case.

nzaugg replied:
Reverted the file-scoped namespacing

legacybass commented on src/EFRepository.Generator/ExtensionMethodGenerator.cs:
```diff
@@ -1,237 +1,388 @@
+			return query;
+
+		if (value != null)
+			return query.Where(n => n.{member.Name}.Date > value.Value.Date);
```
Since the method is called `OnDate`, shouldn't this be an `==` check?

legacybass reviewed (changes requested):
There are a handful of things that should be addressed before we can continue. Some of the comments may not require changes, but others do. Remember that this extension is meant to target code that may not be running the latest .NET version or with the latest features enabled (the single-line namespace is an example of something that would only compile if it were added to a project using language version 10).

natezaugg committed:
PR changes

natezaugg committed:
More changes for the PR
Simplified and organized unit tests.
Added Unit Tests for string IsNull and IsNotNull and IsNotNullOrWhiteSpace...
